File: packages/api-contracts/src/coach.ts
import { z } from 'zod';
import { workoutSetKindSchema } from './workouts';

/**
 * Contrats du coach IA (/api/v1/coach).
 *
 * **L'IA propose, l'application exécute.** Le coach n'écrit rien : sa seule
 * sortie structurée est une proposition de séance, que l'utilisateur accepte
 * explicitement. La création passe ensuite par le chemin de séance existant.
 *
 * Les identifiants de fil et de message sont des UUID générés SUR L'APPAREIL :
 * un message peut être composé avant que le serveur n'ait jamais entendu
 * parler du fil, et l'envoi est rejouable.
 */

export const coachMessageRoleSchema = z.enum(['USER', 'ASSISTANT']);
export type CoachMessageRole = z.infer<typeof coachMessageRoleSchema>;

/**
 * Série proposée — MÊME FORME qu'une série prévue de séance
 * (`WorkoutSessionPlanItem`), volontairement : accepter une proposition est
 * alors une copie, pas une traduction.
 */
export const coachProposalItemSchema = z.object({
  id: z.string(),
  exercisePosition: z.number(),
  /** Toujours un exercice RÉEL du catalogue — vérifié côté serveur. */
  exerciseId: z.string(),
  exerciseName: z.string(),
  setPosition: z.number(),
  kind: workoutSetKindSchema,
  targetReps: z.number().nullable(),
  targetWeightKg: z.number().nullable(),
  restSeconds: z.number().nullable(),
});
export type CoachProposalItem = z.infer<typeof coachProposalItemSchema>;

export const coachSessionProposalSchema = z.object({
  id: z.string(),
  name: z.string(),
  estimatedMinutes: z.number(),
  /** Modèle dont la proposition est dérivée, s'il y en a un. */
  sourceTemplateId: z.string().nullable(),
  /** Séance réellement lancée depuis cette proposition, le cas échéant. */
  acceptedSessionId: z.string().nullable(),
  items: z.array(coachProposalItemSchema),
});
export type CoachSessionProposal = z.infer<typeof coachSessionProposalSchema>;

export const coachMessageSchema = z.object({
  id: z.string(),
  role: coachMessageRoleSchema,
  content: z.string(),
  /** Proposition rattachée au message, quand le coach en a formulé une. */
  proposal: coachSessionProposalSchema.nullable(),
  createdAt: z.string(),
});
export type CoachMessage = z.infer<typeof coachMessageSchema>;

export const coachConversationSummarySchema = z.object({
  id: z.string(),
  title: z.string().nullable(),
  messagesCount: z.number(),
  updatedAt: z.string(),
});
export type CoachConversationSummary = z.infer<typeof coachConversationSummarySchema>;

export const coachConversationSchema = coachConversationSummarySchema.extend({
  messages: z.array(coachMessageSchema),
});
export type CoachConversation = z.infer<typeof coachConversationSchema>;

/** Réponse à un envoi : le message écrit ET la réplique du coach. */
export const coachReplySchema = z.object({
  userMessage: coachMessageSchema,
  assistantMessage: coachMessageSchema,
  /** Messages restants pour la période en cours, après cet échange. */
  remainingToday: z.number(),
});
export type CoachReply = z.infer<typeof coachReplySchema>;

// Les AMORCES de conversation ne sont pas un contrat d'API, et ne l'ont
// jamais été. Un `coachSuggestionsSchema` vivait ici, présenté comme « calculé
// depuis l'état réel de l'utilisateur » côté serveur : aucune route ne le
// servait, aucun client ne le lisait. Le calcul est LOCAL et l'a toujours été
// (`apps/mobile/lib/features/coaching/domain/services/coach_suggestions.dart`,
// branché par `coach_controllers.dart`) — c'est d'ailleurs le bon endroit :
// les amorces dépendent de ce que l'appareil sait déjà, et n'ont pas à coûter
// un aller-retour réseau. Un contrat publié qui décrit un comportement
// inexistant est pire qu'une absence : il se lit comme une promesse.

// ── Requêtes ──────────────────────────────────────────────────────────────

export const createCoachConversationRequestSchema = z.object({
  id: z.string().uuid(),
});
export type CreateCoachConversationRequest = z.infer<typeof createCoachConversationRequestSchema>;

export const sendCoachMessageRequestSchema = z.object({
  /** UUID du message, généré sur l'appareil : l'envoi est rejouable. */
  id: z.string().uuid(),
  content: z.string().min(1).max(2000),
});
export type SendCoachMessageRequest = z.infer<typeof sendCoachMessageRequestSchema>;

export const acceptCoachProposalRequestSchema = z.object({
  /** Séance née sur l'appareil depuis cette proposition. */
  sessionId: z.string().uuid(),
});
export type AcceptCoachProposalRequest = z.infer<typeof acceptCoachProposalRequestSchema>;

File: apps/mobile/lib/features/coaching/domain/entities/coach.dart
/// Entités du domaine coach (immuables, écrites à la main).
library;

import '../../../workout_session/domain/entities/workout.dart';

enum CoachRole {
  user('USER'),
  assistant('ASSISTANT');

  const CoachRole(this.apiValue);

  final String apiValue;

  static CoachRole fromApi(String value) => CoachRole.values.firstWhere(
    (role) => role.apiValue == value,
    orElse: () => CoachRole.assistant,
  );
}

/// Série d'une séance proposée, **telle que le serveur la renvoie** : à plat,
/// une ligne par série, exactement comme le plan d'une séance.
///
/// C'est cette forme-là — et non le regroupement d'affichage — qui permet de
/// lancer la séance : accepter une proposition est alors une copie, pas une
/// traduction.
class CoachProposalSet {
  const CoachProposalSet({
    required this.id,
    required this.exercisePosition,
    required this.exerciseId,
    required this.exerciseName,
    required this.setPosition,
    required this.kind,
    this.targetReps,
    this.targetWeightKg,
    this.restSeconds,
  });

  final String id;
  final int exercisePosition;

  /// Toujours un exercice RÉEL du catalogue : le serveur rejette la
  /// proposition entière si un identifiant lui est inconnu.
  final String exerciseId;
  final String exerciseName;
  final int setPosition;
  final SetKind kind;
  final int? targetReps;
  final double? targetWeightKg;
  final int? restSeconds;
}

/// Exercice d'une séance proposée, **regroupé pour l'affichage**.
///
/// L'API renvoie des séries à plat — une ligne par série, comme
/// `WorkoutSessionPlanItem` — parce que c'est cette forme-là que l'application
/// sait lancer. Le regroupement par exercice appartient à la couche data ; un
/// widget qui le referait à chaque image le referait pour rien.
class CoachProposedExercise {
  const CoachProposedExercise({
    required this.name,
    required this.setCount,
    required this.detail,
  });

  final String name;
  final int setCount;

  /// Résumé des cibles, déjà formaté (« 8 reps · 60 kg »).
  final String detail;
}

/// Séance proposée par le coach. Tant qu'elle n'est pas acceptée, ce n'est
/// qu'un document : elle ne compte ni dans l'historique, ni dans les stats,
/// ni dans les records.
class CoachSessionProposal {
  const CoachSessionProposal({
    required this.id,
    required this.name,
    required this.estimatedMinutes,
    required this.exercises,
    this.sets = const [],
    this.acceptedSessionId,
  });

  final String id;
  final String name;
  final int estimatedMinutes;

  /// Vue d'affichage, regroupée par exercice.
  final List<CoachProposedExercise> exercises;

  /// Vue exécutable, à plat — ce qui devient le plan de la séance.
  final List<CoachProposalSet> sets;

  /// Séance née de cette proposition, quand elle a déjà été lancée.
  final String? acceptedSessionId;

  bool get isAccepted => acceptedSessionId != null;
}

class CoachMessage {
  const CoachMessage({
    required this.id,
    required this.role,
    required this.content,
    this.proposal,
  });

  final String id;
  final CoachRole role;
  final String content;

  /// Proposition rattachée au message, quand le coach en a formulé une.
  final CoachSessionProposal? proposal;
}

/// Fil de discussion complet.
class CoachConversation {
  const CoachConversation({
    required this.id,
    required this.messages,
    this.title,
  });

  final String id;

  /// Du plus ancien au plus récent : c'est l'ordre de lecture, et l'écran
  /// inverse lui-même sa liste pour ancrer la conversation en bas.
  final List<CoachMessage> messages;
  final String? title;
}

/// Résumé d'un fil, pour la liste des conversations.
class CoachConversationSummary {
  const CoachConversationSummary({
    required this.id,
    required this.messagesCount,
    required this.updatedAt,
    this.title,
  });

  final String id;
  final int messagesCount;
  final DateTime updatedAt;
  final String? title;
}

/// Réponse à un envoi : le message écrit ET la réplique du coach.
class CoachReply {
  const CoachReply({
    required this.userMessage,
    required this.assistantMessage,
    required this.remainingToday,
  });

  final CoachMessage userMessage;
  final CoachMessage assistantMessage;

  /// Messages restants pour la journée après cet échange — décidé par le
  /// serveur, jamais compté sur l'appareil.
  final int remainingToday;
}
